### ISSUE TITLE
Explore integration with Dagster

### ISSUE THREAD

**Issue by astrojuanlu** [kedro-org/kedro#3180]
## Description
I have heard from several data people that they're happy with Dagster, which is probably the only "modern", widely used orchestrator that is not mentioned in our docs.

There was a request upstream to add Kedro integration to Dagster https://github.com/dagster-io/dagster/issues/2062 but it's unclear what finally happened.

**Comment by datajoely** [comment]
I think it involves more of spike to work out how it would actually work. I think Flyte (LFAI), Dagster and Metaflow all fall into the modern orchestrator space which isn't served by Kedro. I also would push we address some of the fundamentals outlined in #3094 before doing this.

**Comment by stichbury** [comment]
Thanks! But in that case, it's not a docs ticket so I'll remove the label.

**Comment by astrojuanlu** [comment]
Thanks both - yeah initially I thought about it as a docs ticket (even though the phrasing didn't match) but you're right, this should be a spike first.

And good point @datajoely on looking at Flyte and Metaflow too (let's call them Tier 3), although both have 0.1x times the PyPI downloads of Dagster, so I wouldn't consider them on the same level of adoption. For reference, Dagster and Prefect (Tier 2) have about the same number of downloads, and both have 0.05x times Airflow (Tier 1). Kedro lies between Tier 2 and 3 at the moment.

**Comment by datajoely** [comment]
Aligned - I also think Dagster is closer to Kedro than the others in terms of granularity. In recent years they've really invested in their dbt integration and perhaps we can take inspiration in how they've done that.

**Comment by MatthiasRoels** [comment]
I never explored Dagster as much as I should have, I really like the idea of software defined assets. However, Dagster looks complicated as it has many concepts to understand. Also not sure on how individual task run (especially in a Kubernetes context).

**Comment by astrojuanlu** [comment]
@gtauzin experimenting with Kedro & Dagster! https://github.com/gtauzin/kedro-spaceflights-dagster

**Comment by gtauzin** [comment]
Hey! Thanks @astrojuanlu  for pinging me, it's nice to see some interest for a dagster integration!

It seems to me kedro and dagster are nicely complementary:
- dagster has an asset-driven perspective which pushes you to define nodes in a graph from the perspective of the assets they generate. However, the node function do not have to return the assets or take the one they depend on as inputs. The data assets I/O is left to the user to write. This can be very confusing at first.
- kedro has numerous data connectors and dataset factories which help provide some structure and clarity to complex pipelines. They can be directly of use to help define dagster "asset factories" and remove the trouble of having to handle I/O.

I also feel as @MatthiasRoels that dagster has a lot of concepts. Each of them separately is not necessarily complex, but the way their relate to each other is not always clear to me from the documentation (and the chatbot in there has confused me more than anything else so far).

For example, there are several way of mapping kedro to dagster because dagster has many concepts around generic tasks:
- an op: a task not necessarily associated to an asset;
- a graph of op;
- an asset: which is also an op;
- a multi asset: an op that defines multiple assets;
- an asset graph: a graph of op that ends up defining an asset. 

In practice, to map kedro nodes, I believe multi assets would make sense even in the case of a node that does not have any outputs (and therefore does not define any assets). This is because ops are second-rate citizens in dagster: they do not even appear on their DAG visualization (the global asset lineage) on the UI, but are presented in a form a of a list lost somewhere in a menu. In the case of the spaceflights example, the last node, "evaluate_model_node" does not have any outputs. Defining it as a multi_asset with a corresponding asset that is intangible allows to have it as a part of the asset DAG:

![image](https://github.com/user-attachments/assets/84cf1374-356a-4e88-b29d-8c47e5f6db78)

This small projects is a way for me to deepen my understanding of both kedro and dagster and this is also something I am planning on using for work in the near future. So if you're interested or are also looking into it, don't hesistate to ping me on the kedro slack, I'd be happy to discuss it more.

**Comment by datajoely** [comment]
Super cool work!

**Comment by fdroessler** [comment]
Love seeing this getting some traction! I'm trying to find time to look into kedro + dagster myself but miserably failing at it 😞 

Just adding a few observations here of things I've seen while looking into this. The approaches that dagsters and kedro take to integrate with other frameworks differ slightly. I'll try to add more details as I understand them in the dagster approach because it might be the less familiar and also will help me structure my thoughts. 

Before starting I just want to mention that I don't think either approach is more legitimate than the other just wanted to share what I found so far. Hope this helps others and I'll try to put some code down or iterate on these thoughts as I go along. Also I am sure @deepyaman will soon have a much better view on this 😄 (congrats btw).

# General approaches
**TLDR:** The difference in approaches mostly evolves around whether the logic for converting kedro to dagster is the responsibility of kedro or dagster.

## Kedro approach:
The kedro approach (as far as there is one according to the examples in the documentation) to integrating with an orchestrator is to transpile the kedro primitives into the orchestrator primitives using the general alignment of DAG abstractions between the two frameworks. This often takes the form of a plugin/cli addition in kedro, that when run, creates some artifacts that can be used by the orchestrator. I sometimes think about it as kedro "wrapping" the orchestrator. I think this is what I see @gtauzin doing very nicely in the current kedro-dagster plugin (please correct me if I'm wrong).

## Dagster approach
In contrast it seems like the approach that dagster takes with integrating other authoring frameworks is to wrap them (I have long hunted for a term that describes dbt and kedro nicely and I think authoring framework is currently my favourite). In essence as far as I understand it dagster still uses the wrapped frameworks cli/functions to actually run the code. E.g. for dbt it uses the `dbt build` command of the dbt cli and for dlt it uses the `dlt.run` function. It will use a translator to interpretate the frameworks dag and create a multi asset plus metadata. 

### There are usually the following components:
- asset decorator: This decorator will be the main entry point to parsing and running the wrapped framework. The output can be a multi asset with internal asset dependency (dbt) or a asset/multi_asset without internal asset dependency (dlt)
- Dagster<Framework>Translator: A class that translates from the wrapped framework to dagster entities e.g. asset key, tags, owners etc.
- ConfigurableResource: This runs the actual framework given the current dagster context
- event iterator: Haven't fully understood this part but I think it allows both converting framework execution events into  dagster events for logging and ability to iterate over results

## Random thought on dependencies:
- Using kedro -> dagster we would need a kedro project having dependencies on dagster and kedro-dagster to make things work while in the dagster approach the kedro project could be on its own without any additional dependencies and only the dagster code environment would need a kedro + dagster-kedro dependency.

**Comment by astrojuanlu** [comment]
I'd like to add some color to @fdroessler comment.

> The kedro approach (as far as there is one according to the examples in the documentation) to integrating with an orchestrator is to transpile the kedro primitives into the orchestrator primitives using the general alignment of DAG abstractions between the two frameworks.
> [...] In contrast it seems like the approach that dagster takes with integrating other authoring frameworks is to wrap them (I have long hunted for a term that describes dbt and kedro nicely and I think authoring framework is currently my favourite). In essence as far as I understand it dagster still uses the wrapped frameworks cli/functions to actually run the code. E.g. for dbt it uses the dbt build command of the dbt cli and for dlt it uses the dlt.run function.

Setting aside the pros and cons of the Kedro approach, historically it has made more sense for translating Kedro to "bigger" systems, like Airflow. We've never tried to call dlt or dbt from Kedro so it's unclear whether we'd replicate the same translation approach or try something different.

---

Now, in my opinion, I don't think it would make sense that Kedro were the "driver" of a Dagster project. So either translating Kedro to Dagster or making Dagster drive Kedro would make more sense to me.

**Comment by gtauzin** [comment]
Thanks for initiating this interesting discussion @fdroessler. I probably need some more time to digest it, but it seems to me, the distinction you make between the two approaches makes sense and indeed kedro-dagster has kedro wrap dagster rather than the opposite. I personally did not consider something like a dagster-kedro integration as I felt 
 as @astrojuanlu that kedro is not meant to be a driver although it can drive in some simple cases. That being said, I would be curious to explore the pros and cons of both approaches.

**Comment by fdroessler** [comment]
> Now, in my opinion, I don't think it would make sense that Kedro were the "driver" of a Dagster project. So either translating Kedro to Dagster or making Dagster drive Kedro would make more sense to me.

💯% agreed @astrojuanlu and @gtauzin, I didn't mean to imply that kedro would be the driver so apologies for any misunderstanding. The point I was trying to make and probably poorly was that in one case kedro is initiating the translation of a pipeline into the "bigger" system and in the other case the "bigger" system knows how to "read/execute" a kedro project.

**Comment by deepyaman** [comment]
> (I have long hunted for a term that describes dbt and kedro nicely and I think authoring framework is currently my favourite)

FWIW I personally use "transformation framework", although that is driven more from their position at the "T" in ELT. Authoring framework seems fine if you want to more generically view them as a place where you define some logic.

@fdroessler's understanding is aligned with mine; since Dagster is operating more as your asset-oriented data platform/single pane of glass and not just an orchestrator, it wants to have visibility into each component. In fact, Dagster is probably moving even more in this direction with a new Components initiative; the idea is to make it trivial to drop in your dbt project via the dbt component, dlt pipeline via a dlt component, and perhaps Kedro through an (official or custom) Kedro component. (The Components functionality is very new/being built.)

I think the ideal state is to have a Dagster-Kedro library, and eventually it could even live in https://github.com/dagster-io/dagster/tree/master/python_modules/libraries (however, right now it looks like community-driven integrations often live in their own repo; e.g. this is the SQLMesh integration that Dagster is tracking: https://github.com/opensource-observer/dagster-sqlmesh).

> I personally did not consider something like a dagster-kedro integration as I felt
as @astrojuanlu that kedro is not meant to be a driver although it can drive in some simple cases.

I don't think this is a big problem. Dagster can orchestrate a Kedro pipeline, similar to how it can orchestrate Python code. I think this fits into the view of Kedro as a micro-orchestrator, and, similar to the dbt integration, a `kedro run` command could result in the creation of some multi-asset.

Especially on the final point, I will admit that my knowledge of Dagster is quite tenuous right now, at best; @gtauzin and @fdroessler you all almost certainly have more experience as a user. :) I think @gtauzin's work is a great starting point in that it actually works and is demoable, rather than just being a conceptual discussion. As soon as it's "ready" and docs are updated, I'd love to try and get eyes from some of the people who are very familiar with Dagster integrations—I'm sure they could provide some thoughts and feedback! I imagine, even if want to structure it more like Dagster-Kedro, a lot of the logic/mapping should be the same regardless. Furthermore, I don't actually know the best model to follow from the perspective of creating a new integration, since (1) I haven't yet had the time to really dig into Dagster-dbt and (2) I've heard Dagster-dbt, while very widely used, was also one of the first integrations; some design decisions are there more for legacy and backwards compatibility reasons, so it may not be the ideal template.

Last but not least, Dagster-Kedro is definitely the way to go where Kedro projects are part of your data stack (e.g. you use dlt for ingestion, Kedro for DE + DS, and some dashboarding tool downstream). However, there are undeniably Kedro users who just want to be orchestrate their Kedro projects. Kedro doesn't have a great, ergonomic default way of doing this, so Kedro-Dagster may still end up being an ideal, Kedro-first way to get your project into production. (Ideally, I think it should be easy for Kedro users who know nothing about Dagster to get their project running using Dagster, and it should also be easy to add other components around it as your scope grows.)

**Comment by datajoely** [comment]
Don't have much to add other than - I love that this is getting momentum it feels like a really powerful pattern

**Comment by gtauzin** [comment]
@deepyaman Thanks for the nice answers - including the teasing about dagster Components ! Looking forward to discovering what it is :)

I definitely need to give some more thoughts to that. I do not yet fully understand how dagster integrations are typically built. I'll have a quick look at dagster-dbt.

The way I think about kedro-dagster vs dagster-kedro at the moment is that they would both need to "read"/"translate" a kedro project. I guess the code to do that would be very similar. They differ in how the pipeline is being executed. kedro-dagster wraps kedro nodes into dagster ops and associate any dagster executor to them. From what I understand, dagster-kedro would call `kedro run`. Do you see any other difference?

> Last but not least, Dagster-Kedro is definitely the way to go where Kedro projects are part of your data stack (e.g. you use dlt for ingestion, Kedro for DE + DS, and some dashboarding tool downstream). However, there are undeniably Kedro users who just want to be orchestrate their Kedro projects. Kedro doesn't have a great, ergonomic default way of doing this, so Kedro-Dagster may still end up being an ideal, Kedro-first way to get your project into production. (Ideally, I think it should be easy for Kedro users who know nothing about Dagster to get their project running using Dagster, and it should also be easy to add other components around it as your scope grows.)

Ok let me see if I understand properly:
- If you are kedro users and you are in charge of the orchestration of your pipelines, kedro-dagster is a legitimate option;
- If you are in charge of the data stack in your organization and you have to manage data assets across teams, you use dagster-kedro to include the kedro projects of the DS team.

Makes sense to me. Is this what you meant?

> I think the ideal state is to have a Dagster-Kedro library, and eventually it could even live in https://github.com/dagster-io/dagster/tree/master/python_modules/libraries (however, right now it looks like community-driven integrations often live in their own repo; e.g. this is the SQLMesh integration that Dagster is tracking: https://github.com/opensource-observer/dagster-sqlmesh).

This would be great and I am very happy to support/discuss with anyone willing to work on it.

My impression is that it might be a harder endeavour to write a dagster integration compared to a kedro plugin. The dagster documentation is sometimes incomplete, and mixes new concepts with legacy ones, and so does the dagster codebase and the existing integrations. The dagster slack relies a lot on chatbots to answer questions, which I am not a fan of: they are trained on the imperfect documentation and legacy code, so they often cannot really help and generate massive amount of text which means it is harder to search/learn from other's questions.

**Comment by deepyaman** [comment]
> The way I think about kedro-dagster vs dagster-kedro at the moment is that they would both need to "read"/"translate" a kedro project. I guess the code to do that would be very similar. They differ in how the pipeline is being executed. kedro-dagster wraps kedro nodes into dagster ops and associate any dagster executor to them. From what I understand, dagster-kedro would call `kedro run`. Do you see any other difference?

That seems reasonable. `kedro run` could be not necessarily as a CLI command, but through invoking Python API, too.

> Makes sense to me. Is this what you meant?

Think our understandings are aligned!

> > I think the ideal state is to have a Dagster-Kedro library, and eventually it could even live in https://github.com/dagster-io/dagster/tree/master/python_modules/libraries (however, right now it looks like community-driven integrations often live in their own repo; e.g. this is the SQLMesh integration that Dagster is tracking: https://github.com/opensource-observer/dagster-sqlmesh).
> 
> This would be great and I am very happy to support/discuss with anyone willing to work on it.
> 
> My impression is that it might be a harder endeavour to write a dagster integration compared to a kedro plugin. 

I agree it would be harder. I would also love to get some input from people who work on Dagster integrations on how best to approach this. Regardless, I think your plugin is a critical starting point, since I don't think most people who work on Dagster are familiar with Kedro and how it could integrate.

**Comment by deepyaman** [comment]
> From what I understand, dagster-kedro would call `kedro run`. 

@gtauzin Actually, thinking about this more, maybe Dagster-Kedro could also convert Kedro nodes/create the appropriate Dagster assets based on the data catalog entries and nodes that create them.

Dagster-dbt calls dbt CLI command because Dagster doesnt know how to execute SQL (I think?), but Dagster does know how to run Python code, so it seems feasible.

If that's the case, it might end up very close to what you have for Kedro-Dagster!

**Comment by gtauzin** [comment]
> @gtauzin Actually, thinking about this more, maybe Dagster-Kedro could also convert Kedro nodes/create the appropriate Dagster assets based on the data catalog entries and nodes that create them.
> 
> If that's the case, it might end up very close to what you have for Kedro-Dagster!

Indeed, what kedro-dagster does is translate every part of the kedro project into the corresponding dagster objects and provide the user with a definitions.py to be used as a dragster code location. (non-kedro) Users might modify those as they see fit. One example might be to change the executor of a job, although this will be also possible by filling up the dragster.yml configuration file that kedro-dagster also provides.

> Ok let me see if I understand properly:
> 
> * If you are kedro users and you are in charge of the orchestration of your pipelines, kedro-dagster is a legitimate option;
> * If you are in charge of the data stack in your organization and you have to manage data assets across teams, you use dagster-kedro to include the kedro projects of the DS team.

I was thinking about this distinction we were making. It seemed to me that even if you are a non-kedro user the approach where you use Dagster to manage various assets in a large organisation, having a kedro project fully translated (and modifiable through definitions.py) is a more attractive option than wrapping kedro. In that case, you do not even need to understand kedro at all, you just get your typical Dagster objects.

> Dagster-dbt calls dbt CLI command because Dagster doesnt know how to execute SQL (I think?), but Dagster does know how to run Python code, so it seems feasible.

I do not know much about dbt, but the more I think about it, the less I understand why in the case of kedro one would prefer to use Dagster to run 'kedro run' through the CLI or python API. I feel it is indeed likely they did not have a way to do a dbt-dagster easily.

**Comment by deepyaman** [comment]
I spoke to the person who largely built the dagster-dbt integration the other day to get his rough thoughts on how he thinks such an integration should work, but, in short, there's no single "right" way. Both translating to Python assets in Dagster and wrapping `kedro run` executions are valid, so whatever works. 😀 

Just throwing this here—the dagster-dbt integration supports ephemeral dbt models (which would be like Kedro `MemoryDataset`), so that might be worth examining when get to the nitty-gritty of how to support that.

He also reiterated the easiest way to share this/get more feedback would be a demo eventually, so it's still on my plate to try it out with a simple project (probably will try https://github.com/deepyaman/jaffle-shop, since it will be familiar to people who know more about dbt than Kedro) sooner than later! :)

**Comment by gtauzin** [comment]
Hi everyone :wave:

After over a year, I believe [kedro-dagster](https://github.com/gtauzin/kedro-dagster) is ready to be tried out! A special thanks to @deepyaman  for the early guidance :)

**Main features**
* Lets you map your existing Kedro pipelines into the Dagster orchestrator: auto-generate a `definitions.py` and use a `dagster.yml` in your config to specify jobs, schedules, executors etc.
* Keeps your Kedro hooks, logging and setup intact. In particular, it works seamlessly with [Kedro-mlflow](https://github.com/Galileo-Galilei/kedro-mlflow).
* Experimentally supports Dagster partitions by fanning-out Kedro nodes acting on partitioned datasets.
* The example repo is a full-blown small project showing how it can be wired up. It makes use of dynamic pipelines and also showcases distributed hyperparameter tuning using a new [`optuna.StudyDataset`](https://docs.kedro.org/projects/kedro-datasets/en/latest/api/kedro_datasets_experimental.optuna.StudyDataset.html) experimental dataset.

**Get started**
* Docs: [https://kedro-dagster.readthedocs.io/](https://kedro-dagster.readthedocs.io/)
* Plugin repo: [https://github.com/gtauzin/kedro-dagster](https://github.com/gtauzin/kedro-dagster)
* Example repo: [https://github.com/gtauzin/kedro-dagster-example](https://github.com/gtauzin/kedro-dagster-example)
* Kedro-Viz of the example repo: [https://gtauzin.github.io/kedro-dagster-example/](https://gtauzin.github.io/kedro-dagster-example/)

**I’d love your help & feedback**
It would mean a lot if you could:

* Try it out in one of your Kedro projects
* Spot issues, missing bits or docs gaps
* Share how you’d use it, or ideas for features/improvements
* Reach out if you would like to contribute!

I am looking forward to hearing what you think and how you might use it! :)

---

Also, to come back to the discussion we had on this thread, I do believe that there could be a Dagster integration as well for Dagster users that only wish to use the [Kedro DataCatalog](https://docs.kedro.org/en/1.0.0/catalog-data/data_catalog/). This would translate each Kedro Dataset into IO managers directly usable in a Dagtser project. If there's interest, I am happy to contribute it in the near future! ;)

**Comment by datajoely** [comment]
great work!

**Comment by rashidakanchwala** [comment]
Some of the Kedro team members will try [kedro-dagster](https://github.com/gtauzin/kedro-dagster) in December. I am adding it to our sprint.

**Comment by gtauzin** [comment]
> Some of the Kedro team members will try [kedro-dagster](https://github.com/gtauzin/kedro-dagster) in December. I am adding it to our sprint.

This is great news! Looking forward to hearing what the teams thinks :)

**Comment by ankatiyar** [comment]
Hi @gtauzin 
I quickly tried out the plugin with a demo project with the quick start guide and some minor things I came across -
In the `dg.toml` file generated, there wasn't quotes around the `root_module` so it errored out for me. And the `defs_module` line `defs_module = "kedro_dagster_example"` also errors out saying the module is not in the project - I believe this should be the root module from the project?

I'll do a deeper dive when I have more time but great work! 💯

**Comment by gtauzin** [comment]
@ankatiyar Thank you for trying it out! Indeed, it's really a bummer the dg CLI config file it does not work out-of-the-box. I have created an issue and I will release a fix ASAP :)

**Comment by gtauzin** [comment]
@ankatiyar Fix released with v0.5.1. I now test the quick start scenario comprehensively so that should not happen again 👍 

Thank you! :)

**Comment by ankatiyar** [comment]
Tried it again today and confirm everything works with the latest release! :) Thanks @gtauzin 🚀 
Also love the logo 🫶

**Comment by DimedS** [comment]
Thanks, @gtauzin! I tested this and it works great.

I’d like to clarify one point to make sure I understand the execution model correctly.

When I run:

```bash
kedro dagster dev
```
the plugin starts a local Dagster server and exposes my Kedro project to Dagster by translating it on the fly via the `definitions.py` file generated by `kedro dagster init`.

However, what happens if I already have a Dagster server running (for example, a shared or remote Dagster deployment) and I just want to execute my Kedro project there?

**Comment by gtauzin** [comment]
Thank you for trying it out @DimedS !

The moment you have a `definitions.py`, you have a Dagster project and you may deploy it as a Dagster code location on a remote Dagster instance as any other Dagster project, even if the assets, jobs, etc... have been created by translating their Kedro counterparts. See the [deployment options on the Dagster docs](https://docs.dagster.io/deployment/oss/deployment-options).

In practice I find that there are two interesting ways to handle deployment stages (local, dev, staging, prod) of a Kedro project with Dagster:

1. Through Kedro envs only with each envs being associated to a deployment stage. Then you create one code location per KEDRO_ENV. This is the approach taken by [kedro-dagster-example](https://github.com/gtauzin/kedro-dagster-example). The advantage is that you just have a single git main branch where everything is consistently defined but the problem is that you can't modify deeply a pipeline in `dev` that is working in `prod`. You can still have a pipeline defined for some Kedro environements and not for others like I do [here](https://github.com/gtauzin/kedro-dagster-example/blob/main/src/kedro_dagster_example/pipeline_registry.py) but it's rarely enough to give users the flexibility they need.

2. You can do Dagster branch deployments. In practice that means that each deployment stage is associated to its own branch and to a Kedro environment named the same way as the branch (see the note below to see why it is important). That means that each deployment stage might differ from the others without constraints.

Note that Kedro-Dagster is made for a Kedro project to be deployed as one code location per Kedro environment (for both 1. and 2.). In practice, it means it prefixes any Dagster object name with the name of the Kedro env it was generated from. This means if you have a catalog entry `companies` defined for two different Kedro environments, `dev` and `prod`, and you deploy (using 1. or 2.) the Dagster code locations associated to those environments on the same Dagster instance, you will see a `dev/companies` and a `prod/companies` treated separately by Dagtser. This is because you usually don't want the `dev` pipelines to modify it as it is used by the `prod` pipeline. Keep in mind there are some potential issues there related to lineage as even if datasets used in `dev` and `prod` actually point to the same data, you won't see the `dev` asset having been materialized when the `prod` one has been materialized. This is the reason of existence of the `base` Kedro environment but once a Kedro catalog has been created for a specific Kedro environment, there is no way to know if a dataset was defined in `base` or in that specific Kedro environment.

I use 1. in `kedro-dagster-example` because it is easier to maintain a single branch for an example project, but I would always try and do 2. in a real-life setup.

If you have any other ideas or expectations on how you would like to see a Kedro project being deployed to Dagster instances, I would be very interested to discuss them. Please let me know!
